### PR TITLE
feat(condo): DOMA-10055 added timeouts to batch functions

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**

### CODE REVIEW

**Review by toplenboren — APPROVED** [review]

File: apps/condo/domains/common/utils/serverSchema/index.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ function logTooManyReturnedIfRequired (tooManyReturnedLimitCounters, allObjects,
 
 class GqlWithKnexLoadList {
 
-    constructor ({ listKey, fields, singleRelations = [], multipleRelations = [], where = {}, sortBy = [] }) {
+    constructor ({ listKey, fields, singleRelations = [], multipleRelations = [], where = {}, sortBy = [], timeoutDuration  }) {
```
Comment (sitozzz) [inline apps/condo/domains/common/utils/serverSchema/index.js:47]: Maybe add the default value here?

File: apps/condo/domains/common/utils/serverSchema/index.js [hunk 2]
```diff
@@ -162,49 +180,63 @@ const loadListByChunks = async ({
...
     let maxIterationsCount = Math.ceil(limit / chunkSize)
     let newChunk = []
     let all = []
     let newChunkLength
     let haveWarnedAboutTooManyObjs = false
 
-    do {
-        newChunk = await list.getAll(context, where, { sortBy, first: chunkSize, skip: skip })
-        newChunkLength = newChunk.length
+    let timeoutFlag = { timeoutOccurred: false }
```
Comment (sitozzz) [inline apps/condo/domains/common/utils/serverSchema/index.js:195]: Why not just a boolean flag is used here?

File: apps/condo/domains/common/utils/serverSchema/index.js [hunk 3]
```diff
@@ -162,49 +180,63 @@ const loadListByChunks = async ({
...
+                haveWarnedAboutTooManyObjs = true
+            }
+
+        } while (--maxIterationsCount > 0 && newChunkLength && !timeoutFlag.timeoutOccurred)
+    }
+    
+    await Promise.race([
+        processChunks(),
+        createTimeoutPromise(timeoutDuration),
+    ])
```
Comment (sitozzz) [inline apps/condo/domains/common/utils/serverSchema/index.js:234]: I'm not sure, but can you check if Promise.race method discards the result of the second promise? 
I have been worrying about possible memory leaks and memory cleaning here

**Review by sitozzz — COMMENTED** [review]
Almost ok

File: apps/condo/domains/common/utils/serverSchema/index.js [hunk 4]
```diff
@@ -62,10 +64,20 @@ class GqlWithKnexLoadList {
...
+        let allLength = 0
+
+        const startTime = Date.now()
 
         let maxiterationsCount = 100 // we need some limits - 100K records is more then enough
         do {
+            const now = Date.now()
+
+            if (!conf.DISABLE_CHUNKS_TIMEOUT && now - startTime >= TIMEOUT_DURATION) {
+                throw new Error('Operation timed out')
```
Comment (sitozzz) [inline apps/condo/domains/common/utils/serverSchema/index.js:76]: Maybe add extra information about what operation was cancelled by timeout?

**Review by sitozzz — APPROVED** [review]
Now looks really good to me

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/common/utils/serverSchema/index.js [hunk 5]
```diff
@@ -199,9 +254,20 @@ const loadListByChunks = async ({
             })
             haveWarnedAboutTooManyObjs = true
         }
-
     } while (--maxIterationsCount > 0 && newChunkLength)
 
+    logger.info({
+        msg: 'Return count',
+        functionName: 'loadListByChunks',
+        schemaName: list.key,
```
Comment (Alllex202) [inline apps/condo/domains/common/utils/serverSchema/index.js:262]: Here we get an empty value

**Review by Alllex202 — APPROVED** [review]

**Pull request merged**